libcaf_core/caf/flow/op/retry.test.cpp

```diff
@@ -0,0 +1,197 @@
+        auto input = std::vector{1, 2, 3, 4};
+        make_observable()
+          .from_container(input)
+          .concat(make_observable().fail<int32_t>(sec::invalid_request))
+          .retry(retry_hdl)
+          .do_on_error([err](const error& what) { *err = what; })
+          .for_each([outputs](const int& xs) { outputs->emplace_back(xs); });
+        run_flows();
+        check_eq(outputs->size(), 16u);
+        check_eq(*err, error{sec::invalid_request});
+      }
+    }
+  }
+  GIVEN("an observable that always completes") {
+    WHEN("calling retry") {
+      THEN("the observer completes without retrying") {
+        auto outputs = std::make_shared<std::vector<int>>();
+        auto err = std::make_shared<error>();
+        auto completed = std::make_shared<bool>(false);
+        auto input = std::vector{1, 2, 3, 4};
+        make_observable()
+          .from_container(input)
+          .retry(retry_hdl)
+          .do_on_error([err](const error& what) { *err = what; })
+          .do_on_complete([completed]() { *completed = true; })
+          .for_each([outputs](const int& xs) { outputs->emplace_back(xs); });
+        run_flows();
+        check_eq(outputs->size(), 4u);
+        check_eq(*err, error{}); // no error
+        check(*completed);
+      }
+    }
+  }
+}
+
+SCENARIO("retry can complete flow despite multiple errors") {
+  auto retry_hdl = [](const caf::error& what) {
+    if (what == sec::runtime_error)
+      return true;
+    else
+      return false;
+  };
+  GIVEN("an initialized retry operator") {
+    auto snk = flow::make_passive_observer<int>();
+    auto uut = raw_sub(make_observable().never<int>(), snk->as_observer(),
+                       retry_hdl);
+    snk->request(42);
+    run_flows();
+    WHEN("calling on_error with runtime_error once") {
+      THEN("flow is completed without forwarding error") {
+        uut->on_next(1);
+        uut->on_next(2);
+        run_flows();
+        uut->on_error(sec::runtime_error);
+        run_flows();
+        uut->on_next(1);
+        uut->on_complete();
+        run_flows();
+        check_eq(snk->buf, std::vector<int>{1, 2, 1});
+        check(snk->completed());
+      }
+    }
+    WHEN("calling on_error with runtime_error twice") {
+      THEN(
+        "flow is completed without forwarding error and emitting item stops") {
+        uut->on_next(1);
+        uut->on_next(2);
+        run_flows();
+        uut->on_error(sec::runtime_error);
+        run_flows();
+        uut->on_next(1);
+        uut->on_error(sec::runtime_error);
+        run_flows();
+        uut->on_next(1);
+        uut->on_next(2);
+        uut->on_complete();
+        run_flows();
+        check_eq(snk->buf, std::vector<int>{1, 2, 1, 1, 2});
+        check(snk->completed());
+        uut->on_next(3);
+        uut->on_next(4);
+        uut->on_complete();
+        run_flows();
+        check_eq(snk->buf, std::vector<int>{1, 2, 1, 1, 2});
+        check(snk->completed());
```

Comment:
Calling `on_next` after `on_complete` has nothing to do with the retry logic we are testing here. Discard. Could be its own separate test, but does not belong here.

```suggestion
```

Reply:
I have added a separate test to check that the items are not emitted after the flow is completed.